src/network/networkd-sysctl.c

```diff
@@ -4,18 +4,196 @@
 #include <linux/if.h>
 #include <linux/if_arp.h>
 
+#include "sd-messages.h"
+
 #include "af-list.h"
+#include "cgroup-util.h"
+#include "fd-util.h"
 #include "missing_network.h"
 #include "networkd-link.h"
 #include "networkd-lldp-tx.h"
 #include "networkd-manager.h"
 #include "networkd-ndisc.h"
 #include "networkd-network.h"
 #include "networkd-sysctl.h"
+#include "path-util.h"
 #include "socket-util.h"
 #include "string-table.h"
 #include "sysctl-util.h"
 
+#if HAVE_VMLINUX_H
+
+#include "bpf-link.h"
+
+#include "bpf/sysctl_monitor/sysctl-monitor-skel.h"
+#include "bpf/sysctl_monitor/sysctl-write-event.h"
+
+static struct sysctl_monitor_bpf *sysctl_monitor_bpf_free(struct sysctl_monitor_bpf *obj) {
+        sysctl_monitor_bpf__destroy(obj);
+        return NULL;
+}
+
+static struct ring_buffer *rb_free(struct ring_buffer *rb) {
+        sym_ring_buffer__free(rb);
+        return NULL;
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct sysctl_monitor_bpf *, sysctl_monitor_bpf_free);
+DEFINE_TRIVIAL_CLEANUP_FUNC(struct ring_buffer *, rb_free);
+
+static int sysctl_event_handler(void *ctx, void *data, size_t data_sz) {
+        struct sysctl_write_event *we = ASSERT_PTR(data);
+        Hashmap **sysctl_shadow = ASSERT_PTR(ctx);
+        _cleanup_free_ char *path = NULL;
+        char *value;
+
+        if (we->version != 1)
+                return log_warning_errno(SYNTHETIC_ERRNO(EINVAL),
+                                "Unexpected sysctl event, disabling sysctl monitoring: %d", we->version);
+
+        path = path_join("/proc/sys", we->name);
+        if (!path) {
+                log_oom();
+                return 0;
+        }
+
+        /* If we never managed this handle, ignore it. */
+        value = hashmap_get(*sysctl_shadow, path);
+        if (!value)
+                return 0;
+
+        if (!strneq(value, we->newvalue, sizeof(we->newvalue)))
+                log_struct(LOG_WARNING,
+                           LOG_MESSAGE("Foreign process '%s' changed sysctl '%s' from '%s' to '%s', conflicting with our setting to '%s'",
+                                        we->comm, we->name, we->current, we->newvalue, value),
+                           "MESSAGE_ID=" SD_MESSAGE_SYSCTL_CHANGED_STR);
```

Comment:
please also add structured fields with details, i.e. add:

```c
"COMM=%s", we->comm,
"SYSCTL=%s", we->name,
"OLDVALUE=%s", we->current,
"NEWVLAUE=%s", we->newvalue,
```

Also, please add an entry for this new structured log msg to catalog/systemd.catalog.in briefly explaining what this is about

Reply:
While at it, I've added `OURVALUE=%s` with the value in the shadow

Reply:
This is what I get in the testing VM:
```
# journalctl MESSAGE_ID=9cf56b8baf9546cf9478783a8de42113 -o json |jq '{COMM, SYSCTL, OLDVALUE, NEWVALUE, OURVALUE}'
{
  "COMM": "sysctl",
  "SYSCTL": "net/ipv6/conf/dummy98/accept_ra",
  "OLDVALUE": "0",
  "NEWVALUE": "1",
  "OURVALUE": "0"
}
{
  "COMM": "sysctl",
  "SYSCTL": "net/ipv6/conf/dummy98/mtu",
  "OLDVALUE": "1550",
  "NEWVALUE": "1360",
  "OURVALUE": "1550"
}
{
  "COMM": "sysctl",
  "SYSCTL": "net/ipv4/conf/dummy98/promote_secondaries",
  "OLDVALUE": "1",
  "NEWVALUE": "0",
  "OURVALUE": "1"
}
{
  "COMM": "sysctl",
  "SYSCTL": "net/ipv6/conf/dummy98/proxy_ndp",
  "OLDVALUE": "0",
  "NEWVALUE": "1",
  "OURVALUE": "0"
}
```

Reply:
Added `"COMM_PID=%d", we->pid,` as well